### PR TITLE
Issue #450. First commit. Queue Based Load Leveling

### PR DESCRIPTION
**Pull request by Amarnath510**
I have completed coding, design (UML), README.md. jUnit tests.
This is my first code submission for open source. 
Please review and let me know the issues.

I have one question,
In the README.md file we have to generate the "pumlid". I don't know how to generate the same.
As of now I have left it blank. Let me know how to generate the same.

### CODE REVIEW

**Comment by iluwatar** [comment]
@Amarnath510 the Travis CI build is failing and needs to be fixed first. Looks like there's wrong artifactId in `queue-load-leveling/pom.xml`.

**Comment by Amarnath510** [comment]
@iluwatar 
I have made changes in queue-load-leveling/pom.xml file and pushed.
Should I again do a pull request or my changes are visible to you?

**Comment by iluwatar** [comment]
@Amarnath510 thanks for the quick fix. The changes are applied automatically as you push. Unfortunately now there's a new (Checkstyle) error in the build - see https://travis-ci.org/iluwatar/java-design-patterns/builds/177451598

**Comment by Amarnath510** [comment]
@iluwatar 
I have updated all the checkstyle issues by looking at the error log.
Out of curiosity, how to get the issues produced in my development environment locally, that is in my Eclipse IDE? If there issues are coming in my environment I can fix them there itself. 
FYI, I have given checkstyle.xml as the default one. Just the way it was mentioned [here](https://github.com/iluwatar/java-design-patterns/wiki/12.-IDE-instructions).

**Comment by iluwatar** [comment]
You can reproduce the Checkstyle issues in Eclipse IDE by selecting the project(s) in package/project explorer, right-clicking the project and choosing Checkstyle - Check Code with Checkstyle. I'll add that to the instructions.

**Comment by Amarnath510** [comment]
@iluwatar 
Thanks for the update. 
But the issue now is different. 

> For more details see: /home/travis/build/iluwatar/java-design-patterns/queue-load-leveling/target/pmd.xml -> [Help 1]
> org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.6:check (default) on project queue-load-leveling: You have 1 PMD violation.

Just to make clear what steps I have done,
1) Forked the project and cloned it.
2) Installed Maven in Eclipse (Neon.1a Release (4.6.1))  and Java Verison 1.8.
3) Imported Existing Maven project by pointing to the root folder.
4) Done all the Checkstyle related stuff.
5) Right click on java-design-pattern in the project explorer - new - other - maven (in filer) - Maven Module
6) Create new project, **queue-load-leveling**
7) Created required Java files and UML files and README.md file.
8) Updated pom.xml file with builder/pom.xml file contents and update artifactid to **queu-load-leveling**

Please let me know if I have missed something here.
I am not able to figure out the issue.

**Comment by iluwatar** [comment]
There is `maven-pmd-plugin` configured in the build and it has detected an error in the project. Scan the build output for more details about the PMD error and fix that.

**Comment by Amarnath510** [comment]
@iluwatar 
How can I check this file, 
For more details see: /home/travis/build/iluwatar/java-design-patterns/queue-load-leveling/target/pmd.xml

**Comment by iluwatar** [comment]
If you scroll the build output a bit upwards you will find this line: `PMD Failure: org.queue.load.leveling.Task:29 Rule:UnusedModifier Priority:3 Avoid modifiers which are implied by the context.`

To build locally you just need to do `mvn clean install` on the project root.

**Comment by Amarnath510** [comment]
@iluwatar 
Thanks for pointing out the error. Now I understood the project flow.
I have fixed all the issues. 
Please review my code and let me know if I need to make any changes.

File: queue-load-leveling/src/main/java/org/queue/load/leveling/Message.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  // Set Method for attribute msg.
+  public void setMsg(String msg) {
+    this.msg = msg;
+  }
+  
+  @Override
+  public String toString() {
+    return msg;
+  }
+}
```
Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/Message.java]: I would refactor to make this class immutable. Remove no-args constructor and setter and make `msg` final.

File: queue-load-leveling/src/main/java/org/queue/load/leveling/MessageQueue.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * 
+ * MessageQueue class.
+ * In this class we will create a Blocking Queue and 
+ * submit/retrieve all the messages from it.
+ */
+public class MessageQueue {
+  
+  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
+  
+  private BlockingQueue<Message> blkQueue;
```
Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/MessageQueue.java]: final

File: queue-load-leveling/src/main/java/org/queue/load/leveling/ServiceExecutor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * 
+ *  ServiceExecuotr class.
+ *  This class will pick up Messages one by one from 
+ *  the Blocking Queue and process them.
+ */
+public class ServiceExecutor implements Runnable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
+  
+  private MessageQueue msgQueue;
```
Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/ServiceExecutor.java]: final

File: queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+public class TaskGenerator implements Task, Runnable {
+
+  
+  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
+  
+  // MessageQueue reference using which we will submit our messages.
+  private MessageQueue msgQueue;
+  
+  // Total message count that a TaskGenerator will submit.
+  private int msgCount;
```
Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java]: Make attributes final

File: queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  /**
+   * Each TaskGenerator thread will submit all the messages to the Queue.
+   * After every message submission TaskGenerator thread will sleep for 1 second.
+   */
+  public void run() {
+    try {
+      while (this.msgCount > 0) {
+        String statusMsg = "Message-" + this.msgCount + " submitted by " + Thread.currentThread().getName();
+        Message newMessage = new Message(statusMsg);
+        this.submit(newMessage);
```
Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java]: Message could be created and submitted on one line

File: queue-load-leveling/src/test/java/org/queue/load/leveling/AppTest.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * THE SOFTWARE.
+ */
+package org.queue.load.leveling;
+
+import java.io.IOException;
+
+import org.junit.Test;
+
+/**
+ * Tests that Caching example runs without errors.
```
Comment (iluwatar) [inline queue-load-leveling/src/test/java/org/queue/load/leveling/AppTest.java]: Fix comment

File: queue-load-leveling/src/test/java/org/queue/load/leveling/MessageQueueTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+public class MessageQueueTest {
+
+  @Test
+  public void messageQueueTest() {
+    
+    MessageQueue msgQueue = new MessageQueue();
+    Message msg = new Message("MessageQueue Test");
+    
+    // submit message
+    msgQueue.submitMsg(msg);
```
Comment (iluwatar) [inline queue-load-leveling/src/test/java/org/queue/load/leveling/MessageQueueTest.java]: Message can be created and submitted on one line

File: queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java (hunk 4 above)
Comment (Amarnath510) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java]: I am using `msgCount` to create particular number of messages which I am submitting to MessageQueue. For each message I am reducing the the `msgCount` value. If I make this value final then I can't keep count of message that I am creating. How to solve this issue?

Comment (iluwatar) [inline queue-load-leveling/src/main/java/org/queue/load/leveling/TaskGenerator.java]: It's ok then. Can you resolve the merge conflict, please?

**Comment by Amarnath510** [comment]
@iluwatar 
May be this questions seems to be too silly. Normally when I use TortoiseSVN, if there is a conflict I usually do code update and resolve the conflicts. 
But in Git I don't have any knowledge on how to fix the below conflict. I am worried that I might screw up the code base. Could be please help me with steps about how to resolve this?

merge-conflict

**Comment by iluwatar** [comment]
The conflict should be easy to resolve. See this [guide](https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/) for instructions how to solve this using the command line only.

**Comment by Amarnath510** [comment]
@iluwatar 
I have fixed the conflict. Thanks for the info.
I have made all the necessary changes after review. Please have a look.

**Comment by iluwatar** [comment]
Good job @Amarnath510 :+1: I'm happy to accept your contribution.

**Comment by Amarnath510** [comment]
@iluwatar 
👍 👍 👍 
Thanks a lot. This is my first contribution to open source and I am really excited.
Grateful for your help at every stage. Also will write the required blog.
Excited to pick up the new pattern. Will do that after few days.

**Comment by iluwatar** [comment]
Excellent, I'm looking forward to hearing from you again :smile:

**Pull request merged**